### PR TITLE
[7.x] Add requeue function to jobs

### PR DESCRIPTION
**Pull request by andershagbard**
This pull requests adds a `requeue` method to database and Redis jobs. The idea comes from https://github.com/laravel/ideas/issues/1381

The method does the same as the `release` method, but before putting the job back into queue, it decrements the attempts attribute.

This handles an issue like this, where the job could fail, because the attempts attribute had exceeded the tries attribute.

```php
Redis::throttle('key')->allow(10)->every(60)->then(function () {
    try {
        // API logic
    }
    catch (Exception) {
        // Release the job back into the queue
        return $this->release(10);
    }
}, function () {
    // Could not obtain lock...
    return $this->requeue(10);
});
```

I think we possibly need to rename this method to something more suited. Maybe `isDeletable` ?
https://github.com/laravel/framework/blob/641a7cda8280ecd3035616d4ce6434434b116624/src/Illuminate/Contracts/Queue/Job.php#L64

_Reopen of #30342_

### CODE REVIEW

**Comment by GrahamCampbell** [comment]
Please update the job contract too.

**Comment by themsaid** [comment]
I wonder what could be a user case for this? This seems to be a bad practice in my opinion since it'll keep the job in queue forever.

I'd set a proper `$tries` on my job and use `release()` knowing that I won't get myself into a job that would keep failing forever.

**Comment by andershagbard** [comment]
> I wonder what could be a good use case for this? This seems to be a bad practice in my opinion since it'll keep the job in queue forever.
> 
> I'd set a proper `$tries` on my job and use `release()` knowing that I won't get myself into a situation where I have a job that would keep coming back to the queue forever. Having a few jobs like this could completely block the queue.
> 
> You can also use `retryUntil()` which will basically retry the job until the set time passes no matter how many attempts are incremented. This will make `release()` act like your `requeue()` but with a fail safe.

Obviously, you need to use this function with caution, and not just requeue it on every exception. The use case I mentioned initially is a very good example.

I could also imagine it could be useful if you work with API calls, and your calls are getting throttled. Then I rather use the `requeue` method, and not increase the `attempts`.

**Comment by themsaid** [comment]
@andershagbard You can use `retryUntil` and never worry about the attempts. I just don't see why we need to add a functionality + extra code while we can already achieve what it's supposed to make possible.

**Comment by andershagbard** [comment]
> @andershagbard You can use `retryUntil` and never worry about the attempts. I just don't see why we need to add a functionality + extra code while we can already achieve what it's supposed to make possible.

Imagine you use the Redis throttle method, and your job keeps failing because of some exception, if you use the retryUntil, it will keep failing and fill the queue. It could be a resource heavy job failing, and using server resources.

The comments in the linked ideas issue explains it pretty well.

**Comment by andershagbard** [comment]
> Please update the job contract too.

Updated.

**Comment by GrahamCampbell** [comment]
Please could you rebase your branch against the latest 7.x, to allow the tests to run against the latest code. `git pull --rebase upstream 7.x && git push --force`

File: src/Illuminate/Queue/Jobs/SyncJob.php [hunk 1]
```diff
@@ -49,6 +49,17 @@ public function release($delay = 0)
         parent::release($delay);
     }
 
+    /**
+     * Requeue the job back.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Jobs/SyncJob.php]: One too many spaces here after `int`. Should be exactly 2, but there are 3.

File: src/Illuminate/Queue/Jobs/RedisJob.php [hunk 2]
```diff
@@ -97,6 +97,19 @@ public function release($delay = 0)
         $this->redis->deleteAndRelease($this->queue, $this, $delay);
     }
 
+    /**
+     * Requeue the job.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Jobs/RedisJob.php]: Same, etc.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
I've left some feedback.

**Comment by themsaid** [comment]
@andershagbard 

That's if you `requeue()`, it'll keep the job forever. However, using `retryUntil` it'll keep retrying until a certain time and then fail. I don't really see why you'd risk and use `requeue()` and have hanging jobs in your queue that you can't clear out.

**Comment by andershagbard** [comment]
@themsaid I would not call this an edge case. The problem becomes much more frequent if you use a funnel.

Let's take your example as a case. If the job gets locked out three times in a row, it will run `fail()` on the first exception caught.

**Comment by themsaid** [comment]
I'm just showing an example, you'll need to use a sensible value based on what you need. If you don't want to have any limits then you can clone the job and reset the attempts to `0` and release it back to queue but I wouldn't recommend that ¯\_(ツ)_/¯ .

**Comment by themsaid** [comment]
@taylorotwell explained a use case to me where this might be needed. However, I think we can achieve this without all the changes in that PR. We could update the payload of the job, reset attempts, and then `release()`.

**Comment by andershagbard** [comment]
@themsaid Yes, this is basically what this PR does, but I am not sure if there is a solution which works for all drivers, without adding these methods.

**Comment by themsaid** [comment]
@andershagbard let me tinker with that a bit. Sorry it took me too long to get it 😀

File: src/Illuminate/Contracts/Queue/Job.php [hunk 3]
```diff
@@ -42,6 +42,23 @@ public function release($delay = 0);
      */
     public function isReleased();
 
+    /**
+     * Requeue the job back into the queue.
+     *
+     * Accepts a delay specified in seconds.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Contracts/Queue/Job.php]: One too many spaces

File: src/Illuminate/Queue/InteractsWithQueue.php [hunk 4]
```diff
@@ -61,6 +61,19 @@ public function release($delay = 0)
         }
     }
 
+    /**
+     * Requeue the job.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/InteractsWithQueue.php]: Same

File: src/Illuminate/Queue/Jobs/DatabaseJob.php [hunk 5]
```diff
@@ -56,6 +56,23 @@ public function release($delay = 0)
         return $this->database->release($this->queue, $this->job, $delay);
     }
 
+    /**
+     * Requeue the job.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Jobs/DatabaseJob.php]: Same

File: src/Illuminate/Queue/Jobs/Job.php [hunk 6]
```diff
@@ -119,6 +126,17 @@ public function release($delay = 0)
         $this->released = true;
     }
 
+    /**
+     * Requeue the job back into the queue.
+     *
+     * @param  int   $delay
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/Jobs/Job.php]: Same

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

**Comment by driesvints** [comment]
@GrahamCampbell why did you close this pr?

**Comment by GrahamCampbell** [comment]
I didn't. GitHub seems to have messed up. I rebased it, but it only let met push the old commits and denied access with the new stuff, so it thinks the remote branch is exactly laravel/framework 6.x.

**Comment by GrahamCampbell** [comment]
Now at https://github.com/laravel/framework/pull/31581. Sorry about this!

**Pull request closed without merging**